cross_domain_composability_protocol.md

```diff
@@ -0,0 +1,804 @@
+
+The Shared Publisher (SP) runs a slightly modified version of the SCP protocol:
+1. It sends a `StartCDCP` message to the appropriate sequencers, starts a timer and waits for NSs' `Vote` messages.
+2. If a timeout occurs or if a `Vote(0)` message is received (indicates a failure), it sends a `Decided(0)` message to the NSs and a `NativeDecided(0)` message to the WS, and terminates.
+3. If it receives a `Vote(1)` message from all NSs, it stops the timer (as it not longer will be used) and sends a `NativeDecided(1)` message to the WS, indicating that the NSs are willing to proceed.
+4. Once a `WSDecided(b)` message is received from the WS, it sends a `Decided(b)` message indicating the result to the NSs, and terminates.
+
+Note that, from the above, the SP works as a middle layer between the NSs and WS.
+It indicates the viability results from the NSs to the WS, who is now allowed to communicate with the external rollup for including its transaction.
+Once the WS receives a response, it indicates the result to the SP, who transmits the result back to the NSs.
+
+The Wrapped Sequencer (WS) has the following rules:
+1. Same as NS.
+2. Same as NS.
+3. Whenever a `stagedMailbox.Write` operation is intercepted and fails:
+  - It sends a `Mailbox` message to the counterparty chain sequencer.
+  - It creates a `stagedMailbox.putOutbox` transaction and adds it locally to the transaction list, placing it before the main transaction.
+  - Then, it re-starts the transaction simulation going back to step 2.
+4. Same as NS.
+5. Same as NS.
+6. In case the transaction simulation is successful, it waits for a `NativeDecided` message from the SP.
+7. In case there's a timeout (before a `WSDecided` message has been sent) or if the transaction simulation fails but not due to a mailbox write or read error, it sends a `WSDecided(0)` message to the SP, indicating failure and terminates.
+8. If a `NativeDecided(0)` message is received from the SP, it removes its transaction (and any created `stagedMailbox`) and terminates.
+9. If a `NativeDecided(1)` message is received from the SP and its local transaction simulation is successful, it stops the timer (as it no longer will be used) and sends a special transaction to the ER, which populates the mailbox and executes the transaction atomically.
+10. If the ER transaction fails, it sends a `WSDecided(0)` message to the SP, indicating failure and terminates.
+11. If the ER transaction is successful, it sends a `WSDecided(1)` message to the SP, indicating success and terminates.
+
+The special atomic transaction to the ER, `safe_execute`, allows an atomic execution of the mailbox staging and the main transaction.
+It has the following pseudo-code:
+```solidity
+function safe_execute(stagedInboxMsgs, stagedOutboxMsgs, mainTx) external {
+    // 1. Pre-populate the staged inbox messages
```

Comment:
We should better define the responsibilities and structure of the SP

Structure:
- A committee of operators
- Selected, potentially, via some dPOS protocol
- Each operator has a signing key

Responsibilities
- Validate the simulation (the messages are correct since once we write to an ER we can't un-write it)
- Validate no transitive dependencies
- Sign (each member of the SP committee) that the tx can safely execute